topo/node/ceos/ceos_test.go

```diff
@@ -0,0 +1,267 @@
+					},
+				},
+			},
+		},
+	}
+)
+
+func TestGenerateSelfSigned(t *testing.T) {
+	ki := fake.NewSimpleClientset(&corev1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "pod1",
+		},
+	})
+	reaction := func(action ktest.Action) (handled bool, ret watch.Interface, err error) {
+		f := &fakeWatch{
+			e: []watch.Event{{
+				Object: &corev1.Pod{
+					Status: corev1.PodStatus{
+						Phase: corev1.PodRunning,
+					},
+				},
+			}},
+		}
+		return true, f, nil
+	}
+	ki.PrependWatchReactor("*", reaction)
+	tests := []struct {
+		desc    string
+		wantErr bool
+		ni      node.Interface
+		pb      *topopb.Node
+		spawner func(string, time.Duration, ...expect.Option) (expect.Expecter, <-chan error, error)
+	}{{
+		desc: "no pb",
+	}, {
+		desc: "valid pb",
+		pb:   validPb,
+		ni: &fakeNode{
+			kClient:   ki,
+			namespace: "test",
+		},
+		spawner: fakeSpawner(&fakeExpect{
+			expects: []expects{
+				{resp: "some stuff>"}, // base
+				{resp: "prompt#"},     // enable
+				{resp: "prompt#"},     // key gen
+				{resp: "prompt#"},     // cert gen
+			},
+			sends: []sends{
+				{err: nil}, // enable
+				{err: nil}, // key gen
+				{err: nil}, // cert gen
+			},
+		}),
+	}, {
+		desc: "retry first prompt",
+		pb:   validPb,
+		ni: &fakeNode{
+			kClient:   ki,
+			namespace: "test",
+		},
+		spawner: fakeSpawner(&fakeExpect{
+			expects: []expects{
+				{err: fmt.Errorf("promptErr")},
+				{resp: "prompt>"}, // base
+				{resp: "prompt#"}, // enable
+				{resp: "prompt#"}, // key gen
+				{resp: "prompt#"}, // cert gen
+			},
+			sends: []sends{
+				{err: nil}, // enable
+				{err: nil}, // key gen
+				{err: nil}, // cert gen
+			},
+		}),
+	}, {
+		desc: "retry enable",
+		pb:   validPb,
+		ni: &fakeNode{
+			kClient:   ki,
+			namespace: "test",
+		},
+		spawner: fakeSpawner(&fakeExpect{
+			expects: []expects{
+				{resp: "some stuff>"},          // base
+				{err: fmt.Errorf("enableErr")}, // enable
+				{resp: "prompt>"},              // base
+				{resp: "prompt#"},              // enable
+				{resp: "prompt#"},              // key gen
+				{resp: "prompt#"},              // cert gen
+			},
+			sends: []sends{
+				{err: nil}, // enable (fail)
+				{err: nil}, // enable
+				{err: nil}, // key gen
+				{err: nil}, // cert gen
+			},
+		}),
+	}, {
+		desc: "close err",
+		pb:   validPb,
+		ni: &fakeNode{
+			kClient:   ki,
+			namespace: "test",
+		},
+		spawner: fakeSpawner(&fakeExpect{
+			expects: []expects{
+				{resp: "prompt>"}, // base
+				{resp: "prompt#"}, // enable
+				{resp: "prompt#"}, // key gen
+				{resp: "prompt#"}, // cert gen
+			},
+			sends: []sends{
+				{err: nil}, // enable
+				{err: nil}, // key gen
+				{err: nil}, // cert gen
+			},
+			closeErr: fmt.Errorf("close err"),
+		}),
+		wantErr: true,
+	}}
+	for _, tt := range tests {
+		t.Run(tt.desc, func(t *testing.T) {
+			spawner = tt.spawner
+			mySecond = 0
+			defer func() {
+				spawner = defaultSpawner
+				mySecond = time.Second
+			}()
```

Comment:
Nit:

In the case that spawner or mySecond in ceos.go ever change the restoring of the defaults would have to change too. Doing

defSpawner := spawner
defMySecond := mySecond
defer func() {
    spawner = defSpawner
    mySecond = defMySecond
}
spawner = tt.Spawner
mySecond = 0

avoids the potential future code change at the cost of a few additional lines, up to you

Reply:
ack